src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AzureFunction/AzureFunctionInProcessExecuteWithWatchersAsyncWrapper.cs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020 New Relic, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using NewRelic.Agent.Api;
+using NewRelic.Agent.Extensions.Providers.Wrapper;
+using NewRelic.Reflection;
+
+namespace NewRelic.Providers.Wrapper.AzureFunction;
+
+public class AzureFunctionInProcessExecuteWithWatchersAsyncWrapper : IWrapper
+{
+    private static ConcurrentDictionary<string, InProcessFunctionDetails> _functionDetailsCache = new();
+    private static Func<object, string> _fullNameGetter;
+    private static Func<object, object> _functionDescriptorGetter;
+    private static Func<object, Guid> _idGetter;
+
+    private static bool _loggedDisabledMessage;
+
+    private static bool _coldStart = true;
+    private static bool IsColdStart => _coldStart && !(_coldStart = false);
+
+
+    public bool IsTransactionRequired => false;
+
+    public CanWrapResponse CanWrap(InstrumentedMethodInfo instrumentedMethodInfo)
+    {
+        return new CanWrapResponse(nameof(AzureFunctionInProcessExecuteWithWatchersAsyncWrapper).Equals(instrumentedMethodInfo.RequestedWrapperName));
+    }
+
+    public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall instrumentedMethodCall, IAgent agent, ITransaction transaction)
+    {
+        if (!agent.Configuration.AzureFunctionModeEnabled) // bail early if azure function mode isn't enabled
+        {
+            if (!_loggedDisabledMessage)
+            {
+                agent.Logger.Info("Azure Function mode is not enabled; Azure Functions will not be instrumented.");
+                _loggedDisabledMessage = true;
+            }
+
+            return Delegates.NoOp;
+        }
+
+        object functionInstance = instrumentedMethodCall.MethodCall.MethodArguments[0];
+
+        _functionDescriptorGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(functionInstance.GetType(), "FunctionDescriptor");
+        var functionDescriptor = _functionDescriptorGetter(functionInstance);
+
+        _fullNameGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<string>(functionDescriptor.GetType(), "FullName");
+        string functionClassAndMethodName = _fullNameGetter(functionDescriptor);
+
+        // cache the function details by function name so we only have to reflect on the function once
+        var inProcessFunctionDetails = _functionDetailsCache.GetOrAdd(functionClassAndMethodName, _ => GetInProcessFunctionDetails(functionClassAndMethodName));
+
+        _idGetter ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<Guid>(functionInstance.GetType(), "Id");
+        string invocationId = _idGetter(functionInstance).ToString();
+
+        agent.Logger.Debug($"Instrumenting in-process Azure Function: {inProcessFunctionDetails.FunctionName} / invocation ID {invocationId} / Trigger {inProcessFunctionDetails.Trigger}.");
```

Comment:
We should probably log this at the finest level since it looks like it would happen with every function invocation. We may also want to avoid the string interpolation here so that the string formatting doesn't need to happen unless that logging level is enabled.